datahost-graphql/test/tpximpact/facet_search_test.clj

```diff
@@ -206,3 +208,100 @@ query testQuery {
     (testing "Query without 'datasets' doesn't hang."
       (let [result (h/with-timeout 5000 (h/execute schema query-no-datasets))]
         (is (not= :timeout result))))))
+
+(def query-all-datasets
+  "
+query testQuery {
+  endpoint {
+    catalog {
+      catalog_query {
+        datasets {
+          id title description publisher creator theme
+        }
+      }
+    }
+  }
+}")
+
+
+(def query-facets-with-search-string
+  "
+query testQuery {
+  endpoint {
+    catalog {
+      catalog_query(search_string: \"hmrc\") {
+        datasets {
+          id
+          title
+        }
+        facets {
+          themes {
+            id
+            enabled
+          }
+          creators {
+            id
+            enabled
+          }
+        }
+      }
+    }
+  }
+}")
+
+(def query-facets-with-search-string+constraint
+  "
+{
+  endpoint {
+    catalog {
+      catalog_query(search_string:\"hmrc\"
+                    themes:[\"http://gss-data.org.uk/def/gdp#trade\"]
+      ) {
+        datasets {
+          title
+          creator
+        }
+        facets {
+          creators {
+            enabled
+            id
+          }
+          themes {
+            enabled
+            id
+          }
+        }
+      }
+    }
+  }
+}")
+
+(deftest facets-with-search-string-test
+  (let [schema (h/catql-schema)
+        all-ds (h/result-datasets (h/execute schema query-all-datasets))
+        txt-filtered (->> all-ds
+                          (search/filter-results
+                           {:CatalogSearchResult/search-string "hmrc"})
+                          set)]
+
+    (testing "Facets query with a search string"
+      (let [result (h/execute schema query-facets-with-search-string)
+            by-creator (set/index txt-filtered [:creator])
+            by-theme   (set/index txt-filtered [:theme])]
+        (is (= 0 (count (h/facets-enabled result :publishers)))
+            "we didn't request 'publishers' facet, so there should be zero.")
+        (is (= (count by-creator) (count (h/facets-enabled result :creators))))
+        (is (= (count by-theme) (count (h/facets-enabled result :themes))))))
+    
+    (testing "Facets query with a search string and theme constraint"
+      (let [result (h/execute schema query-facets-with-search-string+constraint)
+            theme "http://gss-data.org.uk/def/gdp#trade"]
+        
+        (testing "facets other than 'theme' should have theme locked"
+          (let [with-locked-theme (filter #(= (:theme %) theme) txt-filtered)
+                num-creators (->> with-locked-theme (group-by :creator) count)]
+            (is (= num-creators (count (h/facets-enabled result :creators))))))
```

Comment:
Logic is the same because, it could be difficult to arrive at the expected result with different logic 😉 You happen to know the answer to this particular query, but how would you verify different queries?